ML Reco becomes friends with Wirecell

bear-is-asleep opened the pull request:
This PR will merge everything from wirecell PR #416  with everything ml reco into fcl files.  See draft PR #420 to see the changes that come just from ml reco. `sbndcode/JobConfigurations/README.md` is further updated to show the standard workflow. For more discussion on changes, see docDB [35186](https://sbn-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=35186) which contains the relevant PRs needed which have all been merged (except one).

There will be a new output of separate `MCParticles` that store particles that are nominally dropped by larg4 (mainly EM shower daughters). We will also store a new product called `SimEnergyDepositLite` which holds a lite version of energy depositions from Geant4. We will also be updating the standard chain to run [supera](https://github.com/DeepLearnPhysics/Supera/tree/develop) and cluster3d by default. 

These changes make new data products and run new stages, and are therefore not considered breaking changes. It will however necessitate slightly new workflows which are outlined in `sbndcode/JobConfigurations/README.md`.

lynnt20 commented on sbndcode/JobConfigurations/standard/detsim/detsim_sce_overlay_lite.fcl:
```diff
@@ -0,0 +1,15 @@
+#include "detsim_sce_lite.fcl"
+
+#Account for track ID offsets in overlay samples by setting mother ID to 0.
+physics.producers.mcreco: @local::sbnd_mcreco_overlay
```
I wouldn't create a new detsim file that will become the standard. I would add this line to the existing `standard_detsim_sbnd.fcl`, so that the names of the "standard" fcls are as consistent as possible. 

I'm not sure that this would affect anyone else who might use mcreco output anyways, so I would just change it in the top-most fcl (which is the standard detsim)

bear-is-asleep replied:
Removed this fcl

lynnt20 commented on sbndcode/JobConfigurations/standard/g4/g4_3drift_faketrigger_filter.fcl:
```diff
@@ -7,6 +7,7 @@ physics.filters.filter: @local::sbnd_larg4faketriggerfilter
 physics.simulate: [ rns
                   , loader
                   , largeant
+                  , sedlite
```
if you're going through all the g4 fcls and adding sedlite, could you also remove all mentions of simdrift and mcreco?

bear-is-asleep replied:
I removed changes to fcls that had simdrift and mcreco

lynnt20 commented on sbndcode/JobConfigurations/README.md:
```diff
@@ -0,0 +1,27 @@
+- `<your-gen>.fcl`
+- `g4_sce_lite.fcl`
+- `detsim_sce_lite.fcl`
+- `reco1_sce_no_overlay_lite.fcl`
```
Similar to my other comment about keeping the standard names as consistent as possible, I'm very hesitant to make the reco/detsim files different for different workflows.. if there is a way to make it the same for rockbox+intime cosmic+single generator, that would be very much preferred.

lynnt20 replied:
@henrylay97 do you have any opinions about the existence of a `reco1_sce_no_overlay_lite.fcl` for non-rockbox, SinglesGen workflow?

henrylay97 replied:
I'll do a full review once Lynn's PR is merged :D But on this I totally agree with Lynn. Is the main difference here due to the `SimpleMerge` module? If so this could be moved right up to `g4` fcl where they are separate fcls anyway?

henrylay97 reviewed (commented):
Hi @bear-is-asleep, thanks for providing the draft PR with isolated changes, very helpful! I agree with @lynnt20 on the idealism of keeping the detsim fcl the same if possible.

Looking at the offsets [here](https://github.com/SBNSoftware/sbndcode/blob/feature/bearc_mlreco_wc_cfg/sbndcode/LArG4/mcreco_sbnd.fcl#L12) and then how they are applied [here](https://github.com/LArSoft/larsim/blob/6ff78c9b83d2afa2ccb4f29130cfaad2d1d64fc5/larsim/MCSTReco/MCRecoPart.cxx#L58-L60) it seems like it would be possible to set the offsets to [0,10000000,20000000] for _all_ samples? Doesn't look (to a naive first glance) like having extra offset values would have any negative consequences?

bear-is-asleep commented:
Hi @lynnt20 and @henrylay97 . Thank you so much for looking these over! I have implemented Henry's suggestion and it's a reasonable change that works. This allowed me to remove the extra detsim fcl. I also removed all instances of `sedlite` for fcls that ran simdrift and mcreco, just to remain consistent with Lynn's changes.

fjnicolas commented:
@bear-is-asleep WireCell PR got merged, I updated `develop`

henrylay97 commented on sbndcode/JobConfigurations/standard/g4/g4_sce_simphotontime_filter.fcl:
```diff
@@ -12,6 +12,13 @@ physics.producers.ionandscintpriorSCE.FillSimEnergyDeposits: true
 physics.producers.ionandscintpriorSCE.InputSourcesLabels: [ "ionandscintintime", "ionandscintouttime"]
 physics.producers.ionandscintpriorSCE.EnergyDepositInstanceLabels: ["priorSCE"]
 
+#Merge dropped MCParticle collections
+physics.producers.largeantdropped: @local::sbnd_merge_dropped_mcpart_overlay
+physics.producers.largeantdropped.InputSourcesLabels: [ "larg4intime:droppedMCParticles", "larg4outtime:droppedMCParticles"]
+
+# Change input sources of simplemerge since we're doing an overlay sample
+physics.producers.simplemerge.InputSourcesLabels: ["largeant", "largeantdropped"]
+
```
This is unnecessary, you've already set these in the `g4_simphotontime_filter.fcl` from which this fcl inherits :smile:

bear-is-asleep replied:
Good catch, thanks!

henrylay97 commented on sbndcode/LArG4/larg4_sbnd.fcl:
```diff
@@ -3,5 +3,8 @@
 BEGIN_PROLOG
 
 sbnd_larg4: @local::standard_larg4
+sbnd_larg4.FillSimEnergyDeposits: true #fill sim energy deposits for ml reco workflow. Will later convert to 
+                                       #sedlite and be dropped from the output
```
What stage are these being dropped in? Not sure I can see it?

bear-is-asleep replied:
Now it is ( :

henrylay97 reviewed (approved):
Looks good Bear. Just one comment for clean up and one that's probably me missing something obvious.

From a workflow point-of-view I'm happy with this now :D 

My approval but should wait on @lynnt20's as well before merging

bear-is-asleep commented:
BTW @fjnicolas this should NOT be merged until this [PR](https://github.com/LArSoft/larg4/pull/52) is merged into larg4.

bear-is-asleep commented:
> BTW @fjnicolas this should NOT be merged until this [PR](https://github.com/LArSoft/larg4/pull/52) is merged into larg4.

It's been merged!

fjnicolas commented:
trigger build

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c14:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/22071&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e26:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/22072&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c14:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13778&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13778&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=build&buildtype=slf7%20c14:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22071&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase unit_test SBND on slf7 for e26:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13779&builds=)

:rotating_light: For more details about the failed phase, check the [unit_test SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13779&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=unit_test&buildtype=slf7%20e26:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22072&builds=)

fjnicolas commented:
trigger build LArSoft/lar*@LARSOFT_SUITE_v09_85_00 SBNSoftware/sbncode#445

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c14:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/22106&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase unit_test SBND on slf7 for c14:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13809&builds=)

:rotating_light: For more details about the failed phase, check the [unit_test SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13809&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=unit_test&buildtype=slf7%20c14:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22106&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e26:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/22105&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e26:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13814&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13814&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=build&buildtype=slf7%20e26:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22105&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase unit_test SBND on slf7 for c14:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13820&builds=)

:rotating_light: For more details about the failed phase, check the [unit_test SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13820&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=unit_test&buildtype=slf7%20c14:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22116&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase ci_tests SBND on slf7 for e26:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13821&builds=)

:rotating_light: For more details about the failed phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13821&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e26:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22118&builds=)

fjnicolas commented:
@bear-is-asleep unit test are now successful. Reco1 is failing though. [Example](https://dbweb8.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/test_details?build_id=sbnd_ci/13821&platform=Linux%203.10.0-1160.105.1.el7.x86_64&buildtype=slf7%20e26:prof&phase=ci_tests&test=nucosmics_reco1_quick_test_sbndcode):

---- StdException BEGIN
37946:         An exception was thrown while processing module LArSoftSuperaDriver/supera run: 1 subRun: 0 event: 1
37947:         [93mMCParticle data pointer not available[00m
37948:       ---- StdException END
37949:       Exception going through path end_path

---- EventProcessorFailure END
37953: %MSG

Probably because the ref files don't have all the necessary info from the MCParticle. I'm running the sequntial tests to check that.

Also, is there a way to reduce the vebosity? It has increased a lot, e.g.
```
1174: ************* Finding best path with A* in cluster *****************
1175: **> There are 136 hits, 96 isolated hits, the alpha parameter is 0.112216
1176: **> PCA len: 4.69472, wid: 0.526821, height: 0.209226, ratio: 0.397149
1177: **> Sorted 96 hits, longest distance: 4.74381
1178: **> Best path has 18 hits, 17 edges
```

bear-is-asleep commented:
> @bear-is-asleep unit test are now successful. Reco1 is failing though. [Example](https://dbweb8.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/test_details?build_id=sbnd_ci/13821&platform=Linux%203.10.0-1160.105.1.el7.x86_64&buildtype=slf7%20e26:prof&phase=ci_tests&test=nucosmics_reco1_quick_test_sbndcode):
> 
> ---- StdException BEGIN 37946: An exception was thrown while processing module LArSoftSuperaDriver/supera run: 1 subRun: 0 event: 1 37947: �[93mMCParticle data pointer not available�[00m 37948: ---- StdException END 37949: Exception going through path end_path
> 
> ---- EventProcessorFailure END 37953: %MSG
> 
> Probably because the ref files don't have all the necessary info from the MCParticle. I'm running the sequntial tests to check that.
> 
> Also, is there a way to reduce the vebosity? There are a lot of couts like:
> 
> ```
> 1174: ************* Finding best path with A* in cluster *****************
> 1175: **> There are 136 hits, 96 isolated hits, the alpha parameter is 0.112216
> 1176: **> PCA len: 4.69472, wid: 0.526821, height: 0.209226, ratio: 0.397149
> 1177: **> Sorted 96 hits, longest distance: 4.74381
> 1178: **> Best path has 18 hits, 17 edges
> ```

I'm going to update the cluster fcl to only run the space point producer, it will also reduce the verbosity to 0!

bear-is-asleep commented:
trigger build

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c14:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/22127&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e26:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/22128&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e26:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13826&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13826&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=build&buildtype=slf7%20e26:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22128&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c14:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/13825&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/13825&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=build&buildtype=slf7%20c14:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/22127&builds=)

fjnicolas commented:
@bear-is-asleep be aware of [3a0b9cb](https://github.com/SBNSoftware/sbndcode/pull/421/commits/3a0b9cbb08198356dabf9b5f403687db5c6f0c62), I needed to added that to make the CI happy

fjnicolas commented:
@bear-is-asleep finally got the CI sequential tests running, see:
https://dbweb8.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/phase_details?build_id=sbnd_ci/13844&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e26:prof
There are a lot of changes as expected, but it ran smoothly. Can you have a look and comment on the changes? Thanks!

is the change in the `rns` at reco1 stage expected?

Note: CPU and memory usage are not up to date (since the WireCell integration), so take the current limits with a grain of salt.

bear-is-asleep commented:
> @bear-is-asleep finally got the CI sequential tests running, see: https://dbweb8.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/phase_details?build_id=sbnd_ci/13844&platform=Linux%203.10.0-1160.105.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e26:prof There are a lot of changes as expected, but it ran smoothly. Can you have a look and comment on the changes? Thanks!
> 
> is the change in the `rns` at reco1 stage expected?
> 
> Note: CPU and memory usage are not up to date (since the WireCell integration), so take the current limits with a grain of salt.

It seems CPU useage has gone up, but we ran some tests and it floats around 5-10 GB of RAM, usually 6 GB for the supera code. This seems to be fine. Execution time will also go up at reco1, adding about 20s for cluster3D and 10s for supera per event. All of the necessary products are in. 

It is a bit strange that `rns` changes, this could be due to adding supera. Is this issue prohibitive to completing the merge? 

Otherwise, the CI tests look good!
